### DISCUSSION TITLE
Limitation compared to babylon js itself?

### DISCUSSION THREAD

**Discussion by TheMetaverseEngineer** [brianzinn/react-babylonjs#287] [Q&A] [answered]
Hi, I have 3 questions:
1. does it have any limitations compared to babylon js?
2.  can we use the new physic engine (havok) now?
3.  can we use it with babylon js editor?

**Comment by brianzinn** [comment, marked as answer]
1.  No. This project is just a renderer. It uses the Babylon.js version of your choice as a peer dependency. 
2. Yes, you’ll need to be on v6 of Babylon.js.
3. Never tried. If you mean using the output of the editor yes - should work.  If you build the gui as react components, you’ll get fast refresh.

**Reply by AmirHo3einJS** [reply]
@brianzinn could you provide an example of using the new physics? Because there is no physicImposter anymore and we must use aggregate!

**Comment by brianzinn** [comment]
You're right @AmirHo3einJS.  Let me add that.  In the meantime, you can use the `onCreated` callback.  Thanks for the great questions - and helping me catch up on v6.  The Physics Aggregate syntax would be like (let me know if that doesn't work):
```jsx
<ground ...>
  <physicsAggregate type={PhysicsShapeType.BOX} _options={{ mass: 0 }} />
</ground>
```

**Reply by AmirHo3einJS** [reply]
You're welcome. thank u for the quick answer

**Reply by pjoe** [reply]
Was `<physicsAggregate>` implemented, don't seem to be in 3.1.20?

**Reply by brianzinn** [reply]
no - it was never added.  It wouldn't take much to add, but I'm super busy now on some projects.  sorry.

**Reply by pjoe** [reply]
Ok, could you point me in the right direction on how/where to add? Happy to attempt a PR 😄

**Reply by brianzinn** [reply]
@pjoe Can you make a new issue or discussion and we can walk through it?  You came across the first hurdle already with IObliqueParams - you can replace that with "any".  That's new in 6.25.0 and I have a PR merged on @babylonjs/core.

You will first need something like a lifecycle listener.
https://github.com/brianzinn/react-babylonjs/blob/master/packages/react-babylonjs/src/customHosts/PhysicsImpostorLifecycleListener.ts

You found likely already the generate code.  The physics aggregate needs to be added there to the `classesToGenerate` list and then we will just need to work backwards from imperative code on the lifecycle listener.

It should be noted that the reconciler will first try to create the Physics Aggregate before what it is attaching to (while what it is attaching to is a constructor argument).  The requested creation order is due to the order that parent/children are created by the reconciler.   So, there is a deferred creation and that we declare on the type itself here:
https://github.com/brianzinn/react-babylonjs/blob/master/packages/react-babylonjs/tools/generate-code.ts#L74

It's a bit convoluted - hoping to refactor that a lot when tree shaking support is added.  Have a look - we can connect on slack or something as well.

I haven't really thought about how specific properties are going to flow back.  We could do an assignFrom for the "body", but even then it has custom methods instead of properties with accessors.  ie: `aggregate.body.setMassProperties({mass: 10});`.  I think that will be a challenge to work out changes on the fly, but setting up should be OK and will fit what I suspect most of use cases.

**Reply by pjoe** [reply]
Already working on a PR, will try to get it up today 😄 Got a bit blocked by #300, but understand the work around.

**Reply by pjoe** [reply]
have things working in #302 😄
